storage/core/src/server/ServerStorage.ts

```diff
@@ -164,6 +165,102 @@ export abstract class ServerStorage
     targetReference: ObjectReference
   ): Promise<void>;
 
+  private async *listObjectsFiltered(
+    sourceStorage: ServerStorage,
+    sourceDirectory: BaseDirectory,
+    maxPageSize: number,
+    predicate?: (objectReference: ObjectReference) => boolean
+  ): AsyncGenerator<ObjectReference> {
+    for await (const objectPage of sourceStorage.getListObjectsPagedIterator(
+      sourceDirectory,
+      maxPageSize
+    )) {
+      for (const object of objectPage) {
+        if (predicate ? predicate(object) : true) {
+          yield object;
+        }
+      }
+    }
+  }
+
+  /**
+   * Copies objects from a base directory in another {@link ServerStorage} instance to this storage.
+   * @param {ServerStorage} sourceStorage source storage. Must be of the same type as this storage.
+   * @param {BaseDirectory} sourceDirectory base directory in the source storage that will be copied.
+   * @param {BaseDirectory} targetDirectory base directory in the target storage.
+   * @param {Function} predicate optional predicate to filter objects to copy. If not specified, all
+   * objects from the sourceDirectory will be copied.
+   * @returns {Promise<void>}
+   * @note This uses server-side copying. Cross-region copy support depends on the storage provider.
+   */
+  public async copyDirectory(
+    sourceStorage: ServerStorage,
+    sourceDirectory: BaseDirectory,
+    targetDirectory: BaseDirectory,
+    predicate?: (objectReference: ObjectReference) => boolean,
+    copyOptions: CopyOptions = {
+      maxPageSize: 100,
+      maxConcurrency: 50,
+      continueOnError: true,
+    }
+  ): Promise<void> {
+    const taskMap = new Map<string, Promise<string>>();
+    const errors: unknown[] = [];
+    const taskFactory = async (object: ObjectReference) => {
```

Comment:
This is quite a complex function, could `taskFactory` and `handleSingleTask` be moved out to be regular private `ServerStorage` functions?

Reply:
I moved `taskFactory` to a separate function while fixing parallelism issues. `handleSingleTask` is there to avoid duplicating complicated code in `for` and `while`, but it's kinda the main part of this function and doesn't really work as a standalone function (unless I wrote a class that would define this task map).